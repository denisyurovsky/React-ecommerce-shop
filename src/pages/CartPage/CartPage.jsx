import CloseIcon from '@mui/icons-material/Close';
import { Button, Modal, Box, Typography } from '@mui/material';
import React, { useState, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';

import Breadcrumbs from '../../components/Breadcrumbs/Breadcrumbs';
import { CartFooter } from '../../components/Cart/CartFooter/CartFooter';
import { CartHeader } from '../../components/Cart/CartHeader/CartHeader';
import { CartProductCards } from '../../components/Cart/CartProductCards/CartProductCards';
import {
  deleteAllProducts,
  deleteProduct,
  getCart,
} from '../../store/cart/cartSlice';
import { getCurrentUser } from '../../store/user/userSlice';

import styles from './CartPage.module.scss';

export const CartPage = () => {
  const dispatch = useDispatch();
  const user = useSelector(getCurrentUser);

  useEffect(() => {
    dispatch(getCart());
  }, [dispatch, user.user.id]);

  const [open, setOpen] = useState(false);
  const [modalProduct, setModalProduct] = useState(undefined);

  const removeAllProducts = () => {
    dispatch(deleteAllProducts());
  };

  const handleConfirmation = () => {
    if (modalProduct) {
      dispatch(deleteProduct({ product: modalProduct }));
    } else {
      removeAllProducts();
    }
    closeModal();
  };

  const closeModal = () => {
    setOpen(false);
    setModalProduct(undefined);
  };

  const openModal = () => {
    setOpen(true);
  };

  const buyHandler = () => {
    alert('Possibility to buy products will be later!');
  };

  return (
    <div className={styles.page}>
      <Breadcrumbs
        links={[
          { text: 'Home', url: '/' },
          { text: 'Cart', url: '' },
        ]}
      />
      <CartHeader openModal={openModal} />
      <CartProductCards
        openModal={openModal}
        setModalProduct={setModalProduct}
      />
      <CartFooter buyHandler={buyHandler} />
      <Modal open={open} onClose={closeModal} className={styles.modal}>
        <Box className={styles.modalInner}>
          <CloseIcon className={styles.closeIcon} onClick={closeModal} />
          <Typography align="center" className={styles.modalText}>
            Are you sure you whant to remove{' '}
            {modalProduct ? modalProduct.name : 'all products'} from the cart?
          </Typography>
          <Box className={styles.modalButtons}>
            <Button
              variant="contained"
              color="primary"
              onClick={handleConfirmation}
            >
              Yes
            </Button>
            <Button variant="outlined" color="primary" onClick={closeModal}>
              No
            </Button>
          </Box>
        </Box>
      </Modal>
    </div>
  );
};
