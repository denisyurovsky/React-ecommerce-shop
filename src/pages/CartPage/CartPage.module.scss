@import "../../assets/styles/colors.scss";

.page {
  width: 100%;
}

.modalButtons {
  width: 30%;
  min-width: 150px;
  margin: auto;
  margin-top: 25px;
  display: flex;
  flex-direction: row;
  justify-content: space-between;
}
