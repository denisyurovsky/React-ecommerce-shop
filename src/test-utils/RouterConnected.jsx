import { createMemoryHistory } from 'history';
import PropTypes from 'prop-types';
import React from 'react';
import { Router, Route, Routes } from 'react-router-dom';

function RouterConnected({ component: Component, path, initialPaths }) {
  const history = createMemoryHistory({ initialEntries: initialPaths });

  return (
    <Router location={history.location} navigator={history}>
      <Routes>
        <Route path={path} element={Component} />
      </Routes>
    </Router>
  );
}

RouterConnected.propTypes = {
  component: PropTypes.object.isRequired,
  path: PropTypes.string,
  initialPaths: PropTypes.arrayOf(PropTypes.string),
};

RouterConnected.defaultProps = {
  path: '/',
  initialPaths: ['/'],
};

export default RouterConnected;
