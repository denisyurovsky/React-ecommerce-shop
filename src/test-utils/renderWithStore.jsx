import { configureStore } from '@reduxjs/toolkit';
import { render as rtlRender } from '@testing-library/react';
import PropTypes from 'prop-types';
import React from 'react';
import { Provider } from 'react-redux';
import { ToastContainer } from 'react-toastify';

import cartReducer from '../store/cart/cartSlice';
import categoriesReducer from '../store/categories/categoriesSlice';
import feedbackReducer from '../store/feedback/feedbackSlice';
import productsReducer from '../store/products/productsSlice';
import userReducer from '../store/user/userSlice';

function render(component, { role = null, store } = {}) {
  let preloadedState;

  if (!store) {
    preloadedState = role
      ? {
          user: {
            user: {
              id: 5,
              role,
            },
          },
        }
      : {};
  }

  const storeWithState = store
    ? store
    : configureStore({
        reducer: {
          cart: cartReducer,
          products: productsReducer,
          categories: categoriesReducer,
          user: userReducer,
          feedback: feedbackReducer,
        },
        preloadedState,
      });

  function Wrapper({ children }) {
    return (
      <Provider store={storeWithState}>
        {children}
        <ToastContainer />
      </Provider>
    );
  }

  Wrapper.propTypes = {
    children: PropTypes.object,
  };

  return rtlRender(component, { wrapper: Wrapper });
}

export * from '@testing-library/react';
export default render;
