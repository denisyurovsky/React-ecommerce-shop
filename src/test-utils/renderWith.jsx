import { configureStore } from '@reduxjs/toolkit';
import { render as rtlRender } from '@testing-library/react';
import PropTypes from 'prop-types';
import React from 'react';
import { Provider } from 'react-redux';
import { ToastContainer } from 'react-toastify';

import cartReducer from '../store/cart/cartSlice';
import categoriesReducer from '../store/categories/categoriesSlice';
import feedbackReducer from '../store/feedback/feedbackSlice';
import productsReducer from '../store/products/productsSlice';
import userReducer from '../store/user/userSlice';

import RouterConnected from './RouterConnected';

function render(
  component,
  preloadedState = {},
  path = '/',
  initialPaths = ['/']
) {
  const store = configureStore({
    reducer: {
      cart: cartReducer,
      products: productsReducer,
      categories: categoriesReducer,
      user: userReducer,
      feedback: feedbackReducer,
    },
    preloadedState,
  });

  function Wrapper({ children }) {
    return (
      <Provider store={store}>
        <RouterConnected
          component={children}
          path={path}
          initialPaths={initialPaths}
        />
        <ToastContainer />
      </Provider>
    );
  }

  Wrapper.propTypes = {
    children: PropTypes.object,
  };

  return rtlRender(component, { wrapper: Wrapper });
}

export * from '@testing-library/react';
export default render;
