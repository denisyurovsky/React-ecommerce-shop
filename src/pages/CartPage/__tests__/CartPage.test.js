import { configureStore } from '@reduxjs/toolkit';
import { fireEvent, render, waitFor, screen } from '@testing-library/react';
import { rest } from 'msw';
import { setupServer } from 'msw/node';
import React from 'react';
import { Provider } from 'react-redux';

import cartReducer from '../../../store/cart/cartSlice';
import userReducer from '../../../store/user/userSlice';
import testProducts from '../../../test-utils/dto/productsDto';
import renderWith from '../../../test-utils/renderWith';
import renderWithStore from '../../../test-utils/renderWithStore';
import RouterConnected from '../../../test-utils/RouterConnected';
import { CartPage } from '../CartPage';

let initialCart = {
  sellers: {
    1: {
      products: [
        {
          userId: 1,
          productId: testProducts[0].id,
          price: testProducts[0].price,
          quantity: 1,
          checked: true,
        },
      ],
      checked: true,
    },
    2: {
      products: [
        {
          userId: 2,
          productId: testProducts[1].id,
          price: testProducts[1].price,
          quantity: 1,
          checked: true,
        },
      ],
      checked: true,
    },
    3: {
      products: [
        {
          userId: 3,
          productId: testProducts[2].id,
          price: testProducts[2].price,
          quantity: 1,
          checked: true,
        },
      ],
      checked: true,
    },
  },
  totalQuantity: 3,
  totalPrice:
    testProducts[0].price + testProducts[1].price + testProducts[2].price,
  isLoading: false,
  errorOccurred: false,
};

let serverUser = {
  id: 1,
  cart: {
    sellers: {
      1: {
        products: [
          {
            userId: 1,
            productId: testProducts[0].id,
            price: testProducts[0].price,
            quantity: 1,
            checked: true,
          },
        ],
        checked: true,
      },
      2: {
        products: [
          {
            userId: 2,
            productId: testProducts[1].id,
            price: testProducts[1].price,
            quantity: 1,
            checked: true,
          },
        ],
        checked: true,
      },
      3: {
        products: [
          {
            userId: 3,
            productId: testProducts[2].id,
            price: testProducts[2].price,
            quantity: 1,
            checked: true,
          },
        ],
        checked: true,
      },
    },
    totalQuantity: 3,
    totalPrice:
      testProducts[0].price + testProducts[1].price + testProducts[2].price,
    isLoading: false,
    errorOccurred: false,
  },
  isLoading: false,
  errorOccurred: false,
};

const initialUser = {
  user: { id: 0 },
};

const stateForTests = {
  user: initialUser,
  cart: initialCart,
};

const store = configureStore({
  reducer: {
    cart: cartReducer,
    user: userReducer,
  },
  preloadedState: stateForTests,
});

const handlersFulfilled = [
  rest.put(`/cart/:userId`, (req, res, ctx) => {
    const { userId } = req.params;

    serverUser = {
      id: userId,
      cart: req.body.cart,
      isLoading: false,
      errorOccurred: false,
    };

    return res(ctx.json({ id: userId, cart: req.body.cart }), ctx.status(200));
  }),
  rest.get('/users/:0', (req, res, ctx) => {
    return res(
      ctx.json({
        ...serverUser,
        firstName: `My firstName`,
        lastName: `My lastName`,
      }),
      ctx.status(200)
    );
  }),
  rest.get('/users/:userId', (req, res, ctx) => {
    const { userId } = req.params;

    return res(
      ctx.json({
        id: userId,
        firstName: `${userId} firstName`,
        lastName: `${userId} lastName`,
      }),
      ctx.status(200)
    );
  }),
  rest.get('/products/', (req, res, ctx) => {
    const response = [];

    if (req.url.searchParams) {
      req.url.searchParams.forEach((item) => {
        response.push(testProducts[item]);
      });
    }

    return res(ctx.json(response), ctx.status(200));
  }),
];

const serverFulfilled = setupServer(...handlersFulfilled);

describe('CartPage component', () => {
  beforeEach(() => serverFulfilled.listen());
  afterEach(() => serverFulfilled.close());

  describe('snapshots', () => {
    it('renders CartPage without data', () => {
      const { asFragment } = renderWithStore(
        <RouterConnected component={<CartPage />} />,
        { store }
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('Operations with products in cart', () => {
    it('should be able to add products', async () => {
      const addRender = renderWith(<CartPage />, stateForTests);
      const title = await addRender.findByText('Incredible Plastic Table');

      await waitFor(() => {
        expect(title).toBeInTheDocument();
      });

      const decreaseButtons = await addRender.findAllByText('-');

      await waitFor(() => {
        expect(decreaseButtons.length).toEqual(3);
      });

      let addButtons = await addRender.findAllByText('+');

      expect(addButtons.length).toEqual(3);
      addButtons = await addRender.findAllByText('+');

      fireEvent.click(addButtons[0]);
      addButtons = await addRender.findAllByText('+');
      expect(screen.getByTestId('SaveIcon')).toBeInTheDocument();

      const quantity = await addRender.findByText('2');

      expect(quantity).toBeInTheDocument();
    });

    it('should be able to decrease products', async () => {
      const decreaseRender = renderWith(<CartPage />, stateForTests);

      const title = await decreaseRender.findByText('Incredible Plastic Table');

      await waitFor(() => {
        expect(title).toBeInTheDocument();
      });

      const decreaseButtons = await decreaseRender.findAllByText('-');

      await waitFor(() => {
        expect(decreaseButtons.length).toEqual(3);
      });

      let secondTitle = await decreaseRender.findByText(
        'Intelligent Cotton Pants'
      );

      expect(secondTitle).toBeInTheDocument();

      fireEvent.click(decreaseButtons[0]);
      expect(screen.getByTestId('SaveIcon')).toBeInTheDocument();
      const testButton = await screen.findAllByText('1');

      expect(testButton).toHaveLength(2);

      secondTitle = await decreaseRender.findByText('Incredible Rubber Cheese');
      const allQuantities = await decreaseRender.findAllByText('1');

      expect(allQuantities.length).toEqual(3);
    });

    it('should be able to delete products', async () => {
      const deleteRender = renderWithStore(
        <RouterConnected component={<CartPage />} />,
        { store }
      );

      const title = await deleteRender.findByText('Incredible Plastic Table');

      await waitFor(() => {
        expect(title).toBeInTheDocument();
      });

      const allQuantities = await deleteRender.findAllByText('1');

      expect(allQuantities.length).toEqual(3);

      const deleteButtons = await deleteRender.findAllByTestId(
        'cartDeleteButton'
      );

      expect(deleteButtons.length).toEqual(3);
      fireEvent.click(deleteButtons[0]);
      expect(await deleteRender.findByText('No')).toBeInTheDocument();
      fireEvent.click(await deleteRender.findByText('No'));
      expect(deleteRender.queryByText('Yes')).toBe(null);
      fireEvent.click(deleteButtons[0]);
      const agreeButton = await deleteRender.findByText('Yes');

      fireEvent.click(agreeButton);

      await waitFor(() => {
        expect(Object.keys(store.getState().cart.sellers).length).toBe(2);
      });
    });

    it('should be able to select products', async () => {
      const selectRender = render(
        <Provider store={store}>
          <RouterConnected component={<CartPage />} />
        </Provider>
      );

      const title = await selectRender.findByText('Incredible Plastic Table');

      await waitFor(() => {
        expect(title).toBeInTheDocument();
      });

      const checkoboxes = await selectRender.findAllByRole('checkbox');

      expect(checkoboxes.length).toEqual(4);
      fireEvent.click(checkoboxes[1]);

      await waitFor(() => {
        expect(store.getState().cart.sellers[2].products[0].checked).toEqual(
          false
        );
      });

      fireEvent.click(checkoboxes[2]);
      await waitFor(() => {
        expect(store.getState().cart.sellers[2].products[0].checked).toEqual(
          false
        );
      });
    });

    it('should be able to delete all products', async () => {
      const deleteAll = renderWith(<CartPage />, stateForTests);

      const deleteAllButton = await deleteAll.findByText('Empty cart');

      fireEvent.click(deleteAllButton);

      expect(await deleteAll.findByText('Yes')).toBeInTheDocument();

      const confirmButton = await deleteAll.findByText('Yes');

      fireEvent.click(confirmButton);

      expect(await deleteAll.findByText('searching')).toBeInTheDocument();
    });
  });
});
