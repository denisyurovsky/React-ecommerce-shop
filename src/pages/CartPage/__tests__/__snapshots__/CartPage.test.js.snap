// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartPage component snapshots renders CartPage without data 1`] = `
<DocumentFragment>
  <div
    class="page"
  >
    <div
      class="breadcrumbs"
      role="presentation"
    >
      <nav
        aria-label="breadcrumb"
        class="MuiTypography-root MuiTypography-body1 MuiBreadcrumbs-root css-63vwl1-MuiTypography-root-MuiBreadcrumbs-root"
      >
        <ol
          class="MuiBreadcrumbs-ol css-4pdmu4-MuiBreadcrumbs-ol"
        >
          <li
            class="MuiBreadcrumbs-li"
          >
            <a
              class="breadcrumb"
              href="/"
            >
              Home
            </a>
          </li>
          <li
            aria-hidden="true"
            class="MuiBreadcrumbs-separator css-1wuw8dw-MuiBreadcrumbs-separator"
          >
            /
          </li>
          <li
            class="MuiBreadcrumbs-li"
          >
            <a
              class="breadcrumb"
              href="/"
            >
              Cart
            </a>
          </li>
        </ol>
      </nav>
    </div>
    <div
      class="header"
    >
      <h2
        class="header"
        data-testid="title"
      >
        Cart
      </h2>
      <button
        class="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root emptyButton css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Empty cart
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
    <div
      class="cards"
    >
      <div>
        <div
          class="names MuiBox-root css-0"
        >
          <span
            class="MuiCheckbox-root MuiCheckbox-colorPrimary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root Mui-checked css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
          >
            <input
              checked=""
              class="PrivateSwitchBase-input css-1m9pwf3"
              data-indeterminate="false"
              type="checkbox"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="CheckBoxIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </span>
          <p
            class="MuiTypography-root MuiTypography-body1 css-ll0ml8-MuiTypography-root"
          >
              
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 css-ll0ml8-MuiTypography-root"
          />
        </div>
        <div
          class="MuiBox-root css-0"
        />
      </div>
      <div>
        <div
          class="names MuiBox-root css-0"
        >
          <span
            class="MuiCheckbox-root MuiCheckbox-colorPrimary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root Mui-checked css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
          >
            <input
              checked=""
              class="PrivateSwitchBase-input css-1m9pwf3"
              data-indeterminate="false"
              type="checkbox"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="CheckBoxIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </span>
          <p
            class="MuiTypography-root MuiTypography-body1 css-ll0ml8-MuiTypography-root"
          >
              
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 css-ll0ml8-MuiTypography-root"
          />
        </div>
        <div
          class="MuiBox-root css-0"
        />
      </div>
      <div>
        <div
          class="names MuiBox-root css-0"
        >
          <span
            class="MuiCheckbox-root MuiCheckbox-colorPrimary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root Mui-checked css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
          >
            <input
              checked=""
              class="PrivateSwitchBase-input css-1m9pwf3"
              data-indeterminate="false"
              type="checkbox"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="CheckBoxIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </span>
          <p
            class="MuiTypography-root MuiTypography-body1 css-ll0ml8-MuiTypography-root"
          >
              
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 css-ll0ml8-MuiTypography-root"
          />
        </div>
        <div
          class="MuiBox-root css-0"
        />
      </div>
    </div>
    <div
      class="footer"
    >
      <div
        class="MuiBox-root css-0"
      >
        <p
          class="totalPrice"
          data-testid="totalPrice"
        >
          Total Price: не число $
        </p>
        <p
          class="savedMoney"
          data-testid="savedMoney"
        >
          You saved: не число $
        </p>
      </div>
      <a
        class="link white uppercase"
        href="/checkout"
        style="opacity: 1;"
      >
        <button
          class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-sghohy-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Buy now
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </a>
    </div>
  </div>
  <div
    class="Toastify"
  />
</DocumentFragment>
`;
